zyte_api/_x402.py

```diff
@@ -0,0 +1,149 @@
+    options = query.get(f"{data_type}Options", {})
+    default_extract_from = "httpResponseBody" if data_type == "serp" else None
+    return options.get("extractFrom", default_extract_from)
+
+
+def get_extract_froms(query: dict[str, Any]) -> set[str]:
+    result = set()
+    for key in EXTRACT_KEYS:
+        if not query.get(key, False):
+            continue
+        result.add(get_extract_from(query, key))
+    return result
+
+
+def may_use_browser(query: dict[str, Any]) -> bool:
+    """Return ``False`` if *query* indicates with certainty that browser
+    rendering will not be used, or ``True`` otherwise."""
+    for key in ("browserHtml", "screenshot"):
+        if query.get(key):
+            return True
+    extract_froms = get_extract_froms(query)
+    if "browserHtml" in extract_froms:
+        return True
+    if "httpResponseBody" in extract_froms:
+        return False
+    return not query.get("httpResponseBody")
+
+
+def _get_eth_key(key: str | None = None) -> str:
+    if key is not None:
+        return key
+    try:
+        return environ[ENV_VARIABLE]
+    except KeyError:
+        raise ValueError from None
+
+
+def get_max_cost_hash(query: dict[str, Any]) -> bytes:
+    """Returns a hash based on *query* that should be the same for queries
+    whose estimate costs are the same.
+
+    For open-ended costs, like actions, network capture or custom attributes,
+    we assume that Zyte API will not report a different cost based on e.g. the
+    number of actions or their parameters, or similar details of network
+    capture or custom attributes.
+
+    See also: https://docs.zyte.com/zyte-api/pricing.html#request-costs
+    """
+    data = {
+        "domain": urlparse(query["url"]).netloc,
+        "type": "browser" if may_use_browser(query) else "http",
+    }
+    for key in (
+        *(k for k in EXTRACT_KEYS if k != "serp"),  # serp does not affect cost
+        "actions",
+        "networkCapture",
+        "screenshot",
+    ):
+        if query.get(key):
+            data[key] = True
+    if query.get("customAttributes"):
+        data["customAttributesOptions.method"] = query.get(
+            "customAttributesOptions", {}
+        ).get("method", "generate")
+    return md5(json.dumps(data, sort_keys=True).encode()).digest()  # noqa: S324
+
+
+async def fetch_requirements(
+    client: x402Client,
+    url: str,
+    query: dict[str, Any],
+    headers: dict[str, str],
+    semaphore: Semaphore,
+    post_fn: Callable[..., AbstractAsyncContextManager[ClientResponse]],
+) -> tuple[Any, str]:
+    post_kwargs = {"url": url, "json": query, "headers": headers}
+    async with semaphore, post_fn(**post_kwargs) as response:
+        if response.status != 402:
+            raise ValueError(
+                "Expected 402 status code for X-402 authorization, got "
+                f"{response.status}"
+            )
+        data = await response.json()
```

Comment:
I wonder if it is OK to keep these requests out of stats, or if we should be logging them in some form, maybe even with its own separate stat just to keep track of the number of such requests.